### ISSUE TITLE
Lb3

### ISSUE THREAD

**Issue by Homka3310** [Homka3310/LearningMachine#4]
import pandas as pd
from gensim.models import KeyedVectors
import numpy as np

# Завантаження даних про столиці та країни
url_capitals = "https://raw.githubusercontent.com/lang-uk/vecs/refs/heads/master/test/test_vocabulary.txt"
data = pd.read_csv(url_capitals, delimiter='\t', header=None, skiprows=1, names=['country1', 'city1', 'country2', 'city2'])

# Завантаження попередньо навченої моделі word2vec
# Попередньо файл треба завантажити та розархівувати
# !wget https://lang.org.ua/static/downloads/models/ubercorpus.cased.tokenized.word2vec.300d.bz2
# !bzip2 -dk ubercorpus.cased.tokenized.word2vec.300d.bz2
model = KeyedVectors.load_word2vec_format("ubercorpus.cased.tokenized.word2vec.300d", binary=False)
import numpy as np

def cosine_similarity(u, v):
  """Обчислює косинусну подібність між векторами u та v."""
  return np.dot(u, v) / (np.linalg.norm(u) * np.linalg.norm(v))

def euclidean_distance(u, v):
  """Обчислює евклідову відстань між векторами u та v."""
  return np.linalg.norm(u - v)
def get_country(city1, country1, city2, word_embeddings):
  """
  Знаходить країну для city2 за аналогією city1 -> country1.
  """
  group = {city1, country1, city2}
  
  # Перевірка наявності слів у моделі
  for word in group:
    if word not in word_embeddings:
      return None, 0.0

  # Векторна арифметика для аналогії
  vec = word_embeddings[country1] - word_embeddings[city1] + word_embeddings[city2]
  
  # Пошук найближчого слова
  # most_similar повертає список (слово, подібність)
  # ми ігноруємо вхідні слова, щоб не знайти їх самих
  closest_words = word_embeddings.most_similar(positive=[vec], topn=10)
  
  for word, similarity in closest_words:
    if word not in group:
      return word, similarity
      
  return None, 0.0
def get_accuracy(word_embeddings, data):
  """
  Обчислює точність моделі аналогій на заданому датасеті.
  """
  num_correct = 0
  
  for i, row in data.iterrows():
    city1, country1, city2, country2 = row['city1'], row['country1'], row['city2'], row['country2']
    
    # Перевіряємо, що всі слова є в моделі
    if all(word in word_embeddings for word in [city1, country1, city2, country2]):
      predicted_country, _ = get_country(city1, country1, city2, word_embeddings)
      
      if predicted_country == country2:
        num_correct += 1
        
  return num_correct / len(data)

# Розрахунок та виведення точності
accuracy = get_accuracy(model, data)
print(f"Точність моделі становить: {accuracy:.2%}")
from sklearn.decomposition import PCA
import matplotlib.pyplot as plt

def visualize(words, model):
  """
  Візуалізує вектори слів у 2D за допомогою PCA.
  """
  # Фільтруємо слова, що є в моделі
  valid_words = [w for w in words if w in model]
  word_vectors = np.array([model[w] for w in valid_words])
  
  # Застосовуємо PCA для зменшення розмірності до 2
  pca = PCA(n_components=2)
  components = pca.fit_transform(word_vectors)
  
  plt.figure(figsize=(12, 8))
  plt.scatter(components[:, 0], components[:, 1])
  
  # Додаємо підписи для кожної точки
  for i, word in enumerate(valid_words):
    plt.annotate(word, xy=(components[i, 0], components[i, 1]))
    
  plt.show()

# Тестування на прикладі
words = ["Україна", "Київ", "Польща", "Варшава", "Німеччина", "Берлін", "футбол", "політика"]
visualize(words, model)
